### ISSUE TITLE
refactor url() into View

### ISSUE THREAD

**Issue by romaninsh** [atk4/ui#307]
Currently to find out target url() we used `$app->url()`. When we are working in context of a view, sometimes we need to know "what's the url that renders this view?". Here is proposed implementation:

``` php
$msg = $tab->add('Message');

$button->link($msg->url());
```

When `View::url()` is called, it would:

 - consult `$this->owner` when unsure.
 - if enclosed in VirtualPage, will add the trigger
 - if enclosed on Tab, add #hash for activating the tab
 - if enclosed on a Loader, also make sure trigger is set
 - if no proper $app is set, throw exception

Some Views can add arguments based on state, for example `Grid` can include page number for #110 to help avoid reliance on session.

Subsequently jsReload() would be refactored to use this `View::url()` method. This also will help developers easily identify which Views cannot be 're-traced' back.

Finally this should help us avoid problems, where beginners are messing up with sticky get.

Cons: slower execution, so perhaps we should think on how to use for TableColumn\Link.

**Comment by DarkSide666** [comment]
StickyGET and URLs as such is a big confusion in UI. So I will vote for having view->url() if that makes it less confusing.
On other hand - I still remember you said that will be bad for security to have view->url() when I proposed such approach back in "toolkit times" :)

**Comment by ibelar** [comment]
If doing so, please make sure that App may have the last word on url somehow and url does not rely entirely on View. As an example, Wordpress integration require to add specific argument in order for ajax to properly run in WP.

```
class AtkWpApp extends App
{
    public function url()
   {
       $this->page = 'admin-ajax';
        if (!WpUtil::isAdmin()) {
            $this->page = WpUtil::getBaseAdminUrl().'admin-ajax';
        }
        $sticky = $this->sticky_get_arguments;
        $sticky['action'] = $this->plugin->getPluginName();
        $sticky['atkwp'] = $this->plugin->getWpComponentId();

        if ($this->plugin->getComponentCount() > 0) {
            $sticky['atkwp-count'] = $this->plugin->getComponentCount();
        }

        if ($this->plugin->config->getConfig('plugin/use_nounce', false)) {
            $sticky['_ajax_nonce'] = helpers\WpUtil::createWpNounce($this->plugin->getPluginName());
        }
  }
}
```

**Comment by romaninsh** [comment]
Here is how i think it would work.

1. each View would have "stickyGetArgs" property = [] by default
2. calling $view->stickyGet() simply adds into that property.
3. adding a child would inherit stickyGetArgs
4. obviously after callbacks some stickiness can also be removed. (stickyForget)
5. calling $view->url() will merge second argument with stickyGetArgs then call $app->url()

So we have a good level of backward compatibility.

WHAT WILL CHANGE?

if any sub-view will create sticky during it's initialization, it won't spill into parent INIT or other objects.

I need some examples for this.

**Comment by acicovic** [comment]
This is merged by #333 so this issue could probably be closed.

**Comment by DarkSide666** [comment]
I would say - close it, but Romans wanted to close issues himself during weekly hangout so adding label.
